### PR TITLE
runtime: pipelined preparation for incoming and delayed receipts

### PR DESCRIPTION
**Pull request by nagisa**
Part of #11319 and the final change in integration with the transaction runtime as all interesting receipt types are handled now. There are also receipt types like yield timeouts which only result in generation of new (delayed) receipts, so they don't need to be handled by this mechanism.

### CODE REVIEW

File: runtime/runtime/src/lib.rs [hunk 1]
```diff
@@ -2150,6 +2217,7 @@ fn missing_chunk_apply_result(
     });
 }
 
+#[inline(never)]
```
Comment (wacban) [inline runtime/runtime/src/lib.rs:2220]: Just out of curiosity, why never inline? If it's non-trivial can you add a comment?

File: runtime/runtime/src/lib.rs [hunk 2]
```diff
@@ -2378,6 +2446,47 @@ struct ApplyProcessingReceiptState<'a> {
...
+
+/// Schedule a one receipt for contract preparation.
+///
+/// The caller should call this method again after the returned number of receipts are processed.
+fn schedule_contract_preparation<'b, R: MaybeRefReceipt>(
+    pipeline_manager: &mut pipelining::ReceiptPreparationPipeline,
+    state_update: &TrieUpdate,
+    mut iterator: impl Iterator<Item = R>,
+) -> Option<usize> {
+    let scheduled_receipt_offset = iterator.position(|peek| {
```
Comment (wacban) [inline runtime/runtime/src/lib.rs:2473]: nit: I think a for loop with an early return would be more readable. I guess it depends on individual preference so up to you.

File: runtime/runtime/src/lib.rs [hunk 3]
```diff
@@ -2378,6 +2446,47 @@ struct ApplyProcessingReceiptState<'a> {
...
+
+impl<'a> MaybeRefReceipt for &'a Receipt {
+    fn as_ref(&self) -> &Receipt {
+        *self
+    }
+}
+
+/// Schedule a one receipt for contract preparation.
+///
+/// The caller should call this method again after the returned number of receipts are processed.
```
Comment (wacban) [inline runtime/runtime/src/lib.rs:2467]: nit: Can you also add what is the return value?

**Review by wacban — APPROVED** [review]
LGTM

File: runtime/runtime/src/lib.rs (hunk 3 above)
Comment (nagisa) [reply runtime/runtime/src/lib.rs:2467]: I'm not sure what is it exactly that you're asking for? Are you asking to document that it is an `Option<usize>`? That's already documented (in the function signature.) The doc-comment itself also explains how to interpret the return value, so I'm really stumped as to what else I could add here :)

File: runtime/runtime/src/lib.rs (hunk 1 above)
Comment (nagisa) [reply runtime/runtime/src/lib.rs:2220]: It is trivial, I've been using it to get more frames into the profiler, it doesn't need to stay.

File: runtime/runtime/src/lib.rs (hunk 3 above)
Comment (wacban) [reply runtime/runtime/src/lib.rs:2467]: Ah sorry I somehow didn't parse the second line of the comment fully.

File: runtime/runtime/src/lib.rs (hunk 1 above)
Comment (nagisa) [reply runtime/runtime/src/lib.rs]: Removed.

File: runtime/runtime/src/lib.rs [hunk 4]
```diff
@@ -1824,6 +1874,20 @@ impl Runtime {
...
+                        // preparation, so lets submit the next one in anticipation that it might
+                        // be processed too (it might also be not if we run out of gas/compute.)
+                        next_schedule_after = schedule_contract_preparation(
+                            &mut processing_state.pipeline_manager,
+                            &processing_state.state_update,
+                            &mut prep_lookahead_iter,
+                        );
+                    }
+                }
+
```
[the remark this answers is not in the document]
Comment (nagisa) [reply runtime/runtime/src/lib.rs:1887]: Yeah, I've been banging my head as to how to best deduplicate this and what I have here is the best I could come up with outside of trying to come up with a way to unify how the receipt types themselves are processed. Perhaps we ought to do that eventually, but it is also non-trivial (esp. due to how different is the handling of delayed receipts.)

**Pull request merged**